### DISCUSSION TITLE
Near/far ranges for plain environment fog

### DISCUSSION THREAD

**Discussion by nenad-jalsovec** [godotengine/godot-proposals#9792] [3D]
The plain environment fog lacks control over fog's near/far ranges. There's only density property. The fog is always applied from the start of the depth range, so even at lowest densities the whole scene catches some fogginess. This is not what you'd typically want with classic fog. There's no way to ensure some minimal "clear" area near the camera. This makes the fog rather unwieldy and pretty much useless for most types of stylized environment rendering.

**Comment by nenad-jalsovec** [comment]
Btw I skimmed the source a bit regarding this and there appear to be some functions that set the range. Not sure why this is not exposed to the editor in WorldEnvironment node or at least to GDScript via RenderingServer singleton.

**Comment by AThousandShips** [comment]
What begin/end are you looking for and how does it differ from [`fog_depth_begin`](https://docs.godotengine.org/en/latest/classes/class_environment.html#class-environment-property-fog-depth-begin)/[`fog_depth_end`](https://docs.godotengine.org/en/latest/classes/class_environment.html#class-environment-property-fog-depth-end)

**Reply by nenad-jalsovec** [reply]
This sounds precisely like what I'm looking for. However I don't see those properties exposed in the inspector for the Environment resource, and I can't find their usage explained anywhere in the docs.

Trying to set `fog_depth_begin` on an Environment resource from a script results in "invalid set index" error. The reference page you linked says it works only if fog_mode is set to FOG_MODE_DEPTH, however FOG_MODE_DEPTH seems to be not defined in the Environment class.

Am I missing something obvious here?

**Reply by AThousandShips** [reply]
They're only available in 4.3

**Reply by nenad-jalsovec** [reply]
Gah! I've spent a week setting up gdextension build setup for 4.2. 
Ok. Looks like I'll have to upgrade when 4.3 is officially released.

Problem solved then I guess. Thanks for the response.

**Reply by AThousandShips** [reply]
Then since this has been implemented this can be closed, ant future improvements to this feature would be handled in separate proposals or reports
